# Files and Tokenization

Using what we have learned in class so far, fork this repository, then:

- Tokenize all of `book_9.txt`. What kinds of errors do you encounter? What other difficulties does this process present? How did you refine your tokenization process through experimentation?
- How many tokens does your tokenization method obtain? How many types? (Remember what tokens and types are?)

In [29]:
from pathlib import Path
import re 

with open("book_9.txt", "r") as file:
    text = file.read()

tokens = text.split()

def refine_tokenization(tokens):
    refined_token = []
    for token in tokens:
        refined_token.extend(re.findall(r"\w+|\W+", token))
    return refined_token

refined_tokens = refine_tokenization(tokens)

refinded_tokens = [token.lower() for token in refined_tokens if token.strip()]

total_tokens = len(refined_tokens)

unique_types = len(set(refined_tokens))
print(f"total tokens: {total_tokens}")
print(f"unique types: {unique_types}")


# out_dir = Path("book_9.txt")

# outfile = out_dir / Path("all_lines.txt")

# if not out_dir.exists():
#     out_dir.mkdir()

total tokens: 8045
unique types: 2007


In [26]:
from pathlib import Path
import re

#opening and reading the file
with open("book_9.txt", "r") as file:
    text = file.read()

#make a list of all the words including regex 
tokens = re.findall(r"[^\s.,!?;:]+", text)

#FAILED ATTEMPT AT TRYING TO EXCLUDE PUNCTUATION AT THE END
#use list comprehension to create a new list and populate it with strip()
# tokens = [token.strip() for token in tokens]
print(f"Amount of Tokens:      {len(tokens)}")
print(f"Amount of Token Types: {len(set(tokens))}")
print(f"Types of Tokens: {set(tokens)}")


Amount of Tokens:      6486
Amount of Token Types: 2046
Types of Tokens: {'rise', 'horrid', 'Juno', 'Rolled', 'public', 'her', 'moved', 'pour', 'leads', 'thought', 'felt', 'promised', 'warn', 'Pluto', 'line', 'deeps', 'his', 'Nestor', 'Or', 'waiting', 'bought', 'fit', 'armies', 'bestowed', 'power', 'gifts', 'Has', 'Venus', 'hands', 'possession', 'inglorious', "time—'twas", 'mix', 'repose', 'unmoved', 'Him', "brother's", 'figure', 'stayed', 'life', 'lofty', 'example', 'left', 'stretched', 'sleepless', 'verge', 'end', 'scorn', 'arms', 'sat', 'Full', 'stratagem', 'Unseen', 'accomplish', 'So', 'march', 'proffered', 'soothes', 'repentance', 'cannot', 'ah', 'vain', 'flamed', 'acclamations', 'once', 'hence', 'reversed', 'Thetis', 'thus', 'anger', 'your', 'contains', 'lovely', 'prowess', 'Cursed', 'means', 'wrath', 'interval', 'thrown', 'implore', 'learn', 'behind', 'ships', 'deserve', 'gentler', 'You', 'both', 'barren', 'Life', "Tydeus'", 'crime', 'withstood', 'majestic', 'Prayers', 'smoking'

### Question 1
#### Grappling w Regex
One of the first challenges was the constructing the correct regex expression, as I wanted tried [\w]+[\W] but the tokens still had punctuation and white space. So I tried \S+ which removed the white space but not the puncuation.
I then resolved to the strip function in a for loop through tokens, but the tokens weren't changing. This is because I was not altering the actual string. As a result, I populated a new lists while looping through the old and simultaneously removing any punctuation. However the punctuation was still there so I changed the regex call to find punctuation and white space first then create the token after.
### Question 2
There are 6486 Tokens and 2046 token types